File: apps/haskell/src/BasicAlgorithms.hs
module BasicAlgorithms
  ( max3
  , mid3
  ) where

-- | 3つの値の最大値を返す
max3 :: Ord a => a -> a -> a -> a
max3 a b c = max a (max b c)

-- | 3つの値の中央値を返す
mid3 :: Ord a => a -> a -> a -> a
mid3 a b c
  | (a >= b && a <= c) || (a >= c && a <= b) = a
  | (b >= a && b <= c) || (b >= c && b <= a) = b
  | otherwise = c

File: apps/elixir/lib/algorithm/basic_algorithms.ex
defmodule Algorithm.BasicAlgorithms do
  @moduledoc """
  基本的なアルゴリズム
  """

  @doc "3 つの値のうち最大値を返す"
  def max3(a, b, c) do
    a |> max(b) |> max(c)
  end

  @doc "3 つの値のうち中央値を返す"
  def mid3(a, b, c) do
    [a, b, c] |> Enum.sort() |> Enum.at(1)
  end
end
